[v1.7] routing: Fix route collisions in AWS ENI (#14337)
cilium-team/cilium: Documentation/install/upgrade.rst

1.7 Upgrade Notes
-----------------

1.7.13 Upgrade Notes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> A new flag ``--egress-multi-home-ip-rule-compat`` was introduced to allow

Reviewer: I wonder if it'd be better to lead with the fix, then describe the flag? EG:

> In rare cases, Cilium 1.7.12 or earlier when configured with ENI IPAM mode
could disrupt connectivity for the node when new ENI devices are created on
the node. This version fixes the issue and introduces a new flag to handle compatibility.

specifying whether to maintain compatibility with the older ENI mode datapath.
It defaults to ``false`` meaning it will instruct Cilium to **not** maintain
compatibility and perform a migration to the new ENI datapath, upon Cilium

> upgrade. This operation will modify the routing table to fix a potential
> collision.

Reviewer: ```suggestion
upgrade. This operation will modify the routing table to fix a collision
that in rare cases can cause connectivity loss to the node.
```

> collision.

Reviewer: As discussed, would be good to have a step here to verify success of the migration.

If the flag is ``true`` upon upgrade, Cilium will **not** perform a migration.

> If the migration was already completed and a downgrade to an older version of
> Cilium is desired, you must first set the flag to ``true`` and then restart
> Cilium so that it can rollback the migration. This ensures that the new ENI
> datapath is rolled back to a state where it is compatible with the older
> version of Cilium.
> 

Reviewer: ```suggestion
If the migration was already completed and you wish to downgrade to
v1.7.12 or earlier, you must first set the flag to ``true`` and then restart
Cilium so that it can rollback the migration. This ensures that the new ENI
datapath is rolled back to a state where it is compatible with the older
version of Cilium. Once this has migrated back to the compatibility mode,
further downgrade to v1.7.12 or earlier can be performed without connectivity
disruption to active pods.

```

.. _current_release_required_changes:

> .. _1.7_required_changes:

Reviewer: nit: Should probably make these link targets above the new section so that users don't accidentally skip over these notes when upgrading.

We can also call this out for ENI users in the release notes.

.. _1.7_required_changes:

1.7.13 Upgrade Notes
~~~~~~~~~~~~~~~~~~~~

In rare cases, Cilium 1.7.12 or earlier when configured with ENI IPAM mode
could disrupt connectivity for the node when new ENI devices are created on the
node. This version fixes the issue and introduces a new flag to handle
compatibility. This flag is ``--egress-multi-home-ip-rule-compat``.

> The flag defaults to ``false`` meaning it will instruct Cilium to perform a
> migration to the new ENI datapath upon Cilium upgrade. This operation will
> modify the routing table to fix a collision that in rare cases can cause
> connectivity loss to the node.

Reviewer: I'm not sure the second sentence adds anything now that the first paragraph explains the bug and what the flag is?

```suggestion
The flag defaults to ``false`` meaning it will instruct Cilium to perform a
migration to the new ENI datapath upon Cilium upgrade. 
```

Author: Good catch, I think this was a leftover from previous feedback and the msg kinda got duplicated.

> The flag defaults to ``false`` meaning it will instruct Cilium to **not**
> maintain compatibility and perform a migration to the new ENI datapath, upon
> Cilium upgrade. This operation will modify the routing table to fix a collision
> that in rare cases can cause connectivity loss to the node.

Reviewer: On second read, `not maintain compatibility` sounds scary if you don't know what's going on. It sounds like there will be network impact. If everything goes well here, there should be no user interaction necessary. I wonder if we can clarify this a bit more.

```suggestion
The flag defaults to ``false`` meaning it will instruct Cilium to perform a
migration to the new ENI datapath upon
Cilium upgrade. This operation will modify the routing table to fix a collision
that in rare cases can cause connectivity loss to the node.
```

If the flag is ``true`` upon upgrade, Cilium will **not** perform a migration.

If the migration was already completed and you wish to downgrade to v1.7.12 or
earlier, you must first set the flag to ``true`` and then restart Cilium so
that it can rollback the migration. This ensures that the new ENI datapath is
rolled back to a state where it is compatible with the older version of Cilium.
Once this has migrated back to the compatibility mode, further downgrade to
v1.7.12 or earlier can be performed without connectivity disruption to active
pods.

> When upgrading to this version, you can verify the success of the migration by
> inspecting for a message like below:

Reviewer: ```suggestion
When upgrading to this version, you can verify the success of the migration by
inspecting Cilium agent logs for a message like below:
```
